notebooks/text_generator.livemd

```diff
@@ -0,0 +1,920 @@
+# Text Generation with LSTM in Axon
+
+```elixir
+Mix.install([
+  {:axon, github: "elixir-nx/axon"},
+  {:nx, "~> 0.2.1"},
+  {:exla, "~> 0.2.2"},
+  {:req, "~> 0.3.0"}
+])
+
+EXLA.set_as_nx_default([:tpu, :cuda, :rocm, :host])
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+:cuda
+```
+
+## Recurrent Neural Networks as generative models
+
+Recurrent neural networks can also be used as generative models.
+
+This means that in addition to being used for predictive models (making predictions) they can learn the sequences of a problem and then generate entirely new plausible sequences for the problem domain.
+
+Generative models like this are useful not only to study how well a model has learned a problem, but to learn more about the problem domain itself.
+
+In this example, we will discover how to create a generative model for text, character-by-character using LSTM recurrent neural networks in Elixir with Axon.
+
+## Preparation
+
+Using [Project Gutenburg](https://www.gutenberg.org/) we can download a text book which is no longer protected under copywrite so we can experiment with them.
+
+The one that we will use for this experiment is [Alice's Adventures in Wonderland by Lewis Carroll](https://www.gutenberg.org/ebooks/11). You can choose any other text or book that you like for this experiment.
+
+```elixir
+filename = "book.txt"
+# change download_url if you like to experiment with other text books.
+download_url = "https://www.gutenberg.org/files/11/11-0.txt"
+
+unless File.exists?(filename) do
+  Req.get!(download_url)
+  |> then(fn %{body: content} -> File.write!(filename, content) end)
+end
+
+book_text = File.read!(filename)
```

Comment:
Since the payload is not really big, we don't need to cache. Also, with this approach changing the URL wouldn't have effect until the file is deleted :)

```suggestion
# Change the URL if you like to experiment with other text books
download_url = "https://www.gutenberg.org/files/11/11-0.txt"

book_text = Req.get!(download_url).body
```